sqlproxy: initial commit and end-to-end demo

tbg opened the pull request:
This adds the proxy envisioned for use with multi-tenant deployments.
Initially, the proxy routes all incoming connections to a backend SQL
server indicated via a command-line flag.

This is good enough for initial sanity checks, but the following changes
need to be made:

- allow `--insecure` (for testing purposes only), which means that some
  of the steps in the proxy are skipped
- add an in-memory Proxy() to (TestServer).StartTenant as well as the
  3node-tenant logic test configuration.
- move `pkg/cmd/mtproxy` to `./cockroach mt start-proxy`. In doing so,
  the proxy will get proper certificate management (the certs types
  will have to be added though), which makes it easier to use.
- implement the parsing of tenant identifier from the database name or
   options field
- load a tenant identifier -> tenant ID map from a file specified via a
    command-line flag
- reload that map on SIGHUP
- unit testing of the resulting code used in the `./cockroach mt
  start-proxy` command (there is already some testing of the `sqlproxy`
  package itself)

The following snippet sets up a demo SQL shell connecting, securely,
through the proxy to the SQL tenant server, which in turn talks to
the KV layer.

    #!/bin/bash

    set -euxo pipefail

    # Prep work
    rm -rf ~/.cockroach-certs cockroach-data
    killall -9 cockroach || true
    killall -9 mtproxy || true

    export CERTSDIR=$HOME/.cockroach-certs
    export COCKROACH_CA_KEY=$CERTSDIR/ca.key
    ./cockroach cert create-ca
    ./cockroach cert create-client root
    ./cockroach cert create-node 127.0.0.1 localhost

    #  Start KV layer. (:26257)
    ./cockroach start --host 127.0.0.1 --background
    # Create tenant
    ./cockroach sql --host 127.0.0.1 -e 'select crdb_internal.create_tenant(123);'

    # Spawn tenant SQL server (:36257) pointing at KV layer (:26257)
    ./cockroach mt start-sql --tenant-id 123 --kv-addrs 127.0.0.1:26257 --sql-addr 127.0.0.1:36257 &
    sleep 1

    # Create user on tenant (proxy does not forward client certs)
    ./cockroach sql --port 36257 -e "create user bob with password 'builder';"

    # Spawn the proxy on :5432, forwarding to tenant SQL server (:36257)
    go run ./pkg/cmd/mtproxy/ -cert-file $CERTSDIR/node.crt -key-file $CERTSDIR/node.key -verify=false --target 127.0.0.1:36257 &

    sleep 5

    # Connect to proxy.
    ./cockroach sql --host 127.0.0.1 --port 5432 --user bob

cc @imjching 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50647)
<!-- Reviewable:end -->

tbg commented:
BTW I'm decidedly lukewarm (actually against it) on step 1) above, but we have committed to an insecure end-to-end demo. Honestly I would be more inclined to just upgrade the demo to being secure, perhaps just using hard-coded certs as done in the demo snippet above - this demonstrates that it just works, now it's just about orchestrating it in a containerized environment. Loading fixture certs into all of the containers is certainly more productive than cluttering the proxy with a special mode that should never be used.

asubiotto reviewed (changes requested):
Agreed re your thoughts on step 1, we should make this demo secure-first. I don't think making it `insecure` buys us anything, and possibly makes the implementation more confusing.

> Reviewed 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MB3jfsY5s4zZJYf3cUb:biqijhm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @nvanbenschoten, and @tbg)

___
*[pkg/cmd/mtproxy/main.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3nJFd1X9P8DKSlvyl:-MB3nJFd1X9P8DKSlvym:b-heprb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L70)):*
> ```Go
> 		IncomingTLSConfig: &tls.Config{Certificates: []tls.Certificate{cer}},
> 		OutgoingTLSConfig: &tls.Config{InsecureSkipVerify: !options.verify},
> 		OutgoingAddrFromParams: func(map[string]string) (addr string, clientErr error) {
> ```

I'm not sure I like passing the proxying logic as a callback. It might be better to define an interface in `sqlproxy` that is implemented here in some `TenantProxier` struct that is passed in to `Serve`. Makes for better unit testing too I think.
___
*[pkg/sql/sqlproxy/error.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qdI_7dgeiXx8wgRD:-MB3qdI_7dgeiXx8wgRE:b-bpd7dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/error.go#L32)):*
> ```Go
> 	// SSLRequest was expected but the client attempted to go ahead without TLS,
> 	// or vice versa.
> 	CodeInsecureUnexpectedStartupMessage
> ```

nit: `s/CodeInsecureUnexpected/CodeUnexpectedInsecure`?
___
*[pkg/sql/sqlproxy/error.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3obUs-i8kPoY0xfdg:-MB3obUs-i8kPoY0xfdh:blmsa5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/error.go#L40)):*
> ```Go
> 	// CodeSecureStartupMessageFailed indicates an unexpected startup message
> 	// received from the client after TLS negotiation.
> 	CodeSecureStartupMessageFailed
> ```

nit: `s/CodeSecureStartupMessageFailred/CodeUnexpectedStartupMessage`?
___
*[pkg/sql/sqlproxy/proxy.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3q3pK6yINmY1yHyOU:-MB3q3pK6yINmY1yHyOV:b-dtzaol) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L33)):*
> ```Go
> 	OnSendErrToClient func(code ErrorCode, msg string) string
> 
> 	_ struct{} // force explicit init of this struct
> ```

OOC, what does this prevent against?
___
*[pkg/sql/sqlproxy/proxy.go, line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qwku41Hp2z2VbVpd:-MB3qwkvANhIkGzmdM8L:bpqz41z) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L37)):*
> ```Go
> 
> func Proxy(conn net.Conn, opts Options) error {
> 	sendErrToClient := func(conn net.Conn, code ErrorCode, msg string) {
> ```

Should we move this out of `Proxy`?
___
*[pkg/sql/sqlproxy/proxy.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3rhFZ5-UefutOCHIx:-MB3rhFZ5-UefutOCHIy:b-2xt8pk) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L49)):*
> ```Go
> 
> 	{
> 		m, err := pgproto3.NewBackend(pgproto3.NewChunkReader(conn), conn).ReceiveStartupMessage()
> ```

I would extract this block into a `initSecureConn` (or similarly named) function
___
*[pkg/sql/sqlproxy/proxy.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t73c7IyWJj1wahOr:-MB3t73c7IyWJj1wahOs:b3jc80o) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L60)):*
> ```Go
> 		}
> 
> 		_, err = conn.Write([]byte("S"))
> ```

Extract `S` into const and reuse below.
___
*[pkg/sql/sqlproxy/proxy.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t29T0ml7Mo79Sh_h:-MB3t29T0ml7Mo79Sh_i:b-oicwr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L109)):*
> ```Go
> 
> 	// Send SSLRequest.
> 	if err := binary.Write(crdbConn, binary.BigEndian, []int32{8, 80877103}); err != nil {
> ```

I think it's worth extracting these into `const`s
___
*[pkg/sql/sqlproxy/proxy_test.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3wUCR22uSA_JISZn1:-MB3wUCR22uSA_JISZn2:brkjub5) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L88)):*
> ```Go
> }
> 
> func assertConnectErr(t *testing.T, prefix, suffix string, expCode ErrorCode, expErr string) {
> ```

expCode is unused
___
*[pkg/sql/sqlproxy/proxy_test.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3wfOb8SfJgnwdZFRS:-MB3wfOb8SfJgnwdZFRT:b-wtmrwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L88)):*
> ```Go
> }
> 
> func assertConnectErr(t *testing.T, prefix, suffix string, expCode ErrorCode, expErr string) {
> ```

call `t.Helper` here?
___
*[pkg/sql/sqlproxy/proxy_test.go, line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qNSb5IWzccDI24qn:-MB3qNSb5IWzccDI24qo:bggaw4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L178)):*
> ```Go
> 	t.Skip("this test needs a running (secure) CockroachDB instance at the given address")
> 	const crdbSQL = "127.0.0.1:52966"
> 	// TODO(tbg): use an in-mem test server once this code lives in the CRDB repo.
> ```

Is that now?
___
*[pkg/sql/sqlproxy/server.go, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3us_p5l87XE-HYjJr:-MB3us_p5l87XE-HYjJs:buax789) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/server.go#L10)):*
> ```Go
> 
> func Serve(ln net.Listener, opts Options) error {
> 	doneCh := make(chan error)
> ```

Where are these channels used?
___
*[pkg/storage/enginepb/mvcc3.pb.go, line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3uLfJ8XfY0QlQ3Df4:-MB3uLfJ8XfY0QlQ3Df5:b-49k3gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/storage/enginepb/mvcc3.pb.go#L0)):*
This is related to the go mod/1.14 bump, right? Don't think this should be checked in.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 7 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @nvanbenschoten, and @tbg)

___
*[pkg/sql/sqlproxy/proxy.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB4nJCZ-zGE0o6YLDHi:-MB4nJCZ-zGE0o6YLDHj:bq0sbmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L21)):*
> ```Go
> 
> type Options struct {
> 	IncomingTLSConfig *tls.Config
> ```

Give these small comments like `// from SQL client` and `// to SQL gateway` so make the "incoming" and "outgoing" more clear.
___
*[pkg/sql/sqlproxy/proxy.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3q3pK6yINmY1yHyOU:-MB4mogh8zsivINA8yV7:bcgkva7) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L33)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

OOC, what does this prevent against?
</blockquote></details>

Yeah, I'm not sure either. I thought maybe it had something to do with mandating initialization style, but that doesn't appear to be the case: https://play.golang.org/p/MbfbbR-2uHW.
___
*[pkg/sql/sqlproxy/proxy.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t29T0ml7Mo79Sh_h:-MB4q_66BfV3GJYJpCpb:bijr8wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L109)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it's worth extracting these into `const`s
</blockquote></details>

+1. Plus we save the allocations if we do so.
___
*[pkg/sql/sqlproxy/proxy.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB4r6Fk5dl4Bvf8J1hp:-MB4r6Fl8VkXtzpNRMAs:bv0aidb) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L128)):*
> ```Go
> 	}
> 
> 	errOutgoing := make(chan error)
> ```

Do these need to be buffered to avoid leaking the goroutine that hits an error second?
___
*[pkg/sql/sqlproxy/server.go, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3us_p5l87XE-HYjJr:-MB4jmlx4ULw7RRrlTQo:b-2si8jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/server.go#L10)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Where are these channels used?
</blockquote></details>

Looks like nowhere. They can likely be removed.


<!-- Sent from Reviewable.io -->

imjching reviewed (commented):
To add on to the list of changes, I think it would be nice to add some sort of liveness checks in the future to know whether the proxy can accept more connections.

> Reviewed 36 of 36 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MBAO8b5CCBEdXs0zPE0:b-o7i3zf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @nvanbenschoten, and @tbg)

___
*[pkg/cmd/mtproxy/main.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBAKnAo38ur45WfqNKW:-MBAKnAo38ur45WfqNKX:b-tzu2kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L52)):*
> ```Go
> 		"Address to proxy to (a Postgres-compatible server)")
> 	flag.BoolVar(&options.verify, "verify", true,
> 		"If true, use InsecureSkipVerify=true for connections to target")
> ```

Did you mean `s/InsecureSkipVerify=true/InsecureSkipVerify=false` here?
___
*[pkg/cmd/mtproxy/main.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBAKvKy6Kut7eQTWDUg:-MBAKvKy6Kut7eQTWDUh:brpj3oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L69)):*
> ```Go
> 	opts := proxy.Options{
> 		IncomingTLSConfig: &tls.Config{Certificates: []tls.Certificate{cer}},
> 		OutgoingTLSConfig: &tls.Config{InsecureSkipVerify: !options.verify},
> ```

Setting `options.verify=true` won't work here as it requires a `ServerName` to be specified. I suppose in the long-term, we would want to set `options.verify=true` in production.
___
*[pkg/cmd/mtproxy/main.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBALRpG2aTQ6SxLYi96:-MBALRpG2aTQ6SxLYi97:bg1nj7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L75)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 			// TODO(asubiotto): implement the actual translation here once it is clear
> 			// how this will work. It's likely that a filename will be passed to the
> 			// proxy which contains a lookup map (and which needs to be re-read on
> 			// SIGHUP). For now, just send everybody to one address and don't validate
> 			// any parameters.
> ```
</details>


We can talk about this when you're back, but it would be nice to implement custom routing policies here. I'm looking for a substitution-based routing policy for free tier: https://github.com/imjching/cockroach/commit/545cf62ad7887546c3a7c4489221d07faccef03e#diff-461d7bf3b3646351a9bcb07c0e43b9b5R109-R110. Since the actual translation (i.e. parsing tenant name) is free-tier specific, I wonder if we should allow users to customize what gets extracted. Alternatively, we could store the tenant name in `systems.tenant` as a CRDB thing, so the parsing logic here would make sense.

One of the reasons why I chose routing-based on DNS names is that it reduces a lot of operational work. Using a map here means we need to keep the map up-to-date whenever we modify the tenant list.
___
*[pkg/sql/sqlproxy/proxy.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB4r6Fk5dl4Bvf8J1hp:-MBAMf4A7jKbe5Q7zDgI:b-ba6x5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L128)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do these need to be buffered to avoid leaking the goroutine that hits an error second?
</blockquote></details>

I'm curious, why would this be the case? The goroutines will only run once each, and therefore we will only write to the channels once each.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Thanks for the reviews all!

Rebased and added all code changes as follow-up commits for reviewing ease (will squash appropriately up before merge). I'm sure the linters will find something to complain so there's at least another pass here, but all comments were addressed (or at least responded to).

@imjching good point, filed https://github.com/cockroachdb/cockroach/issues/51081.

@otan can you look at the (most recent) commit with the vendor change? It's unfortunate that I'm seeing honnef-tools bumped, is that going to wreak havoc? Also there seem to be two pgx imports (/v4 and v3.6.2), is that benign?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MBehbZm0oySJp1YlgwU:bo7te3o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @imjching, and @nvanbenschoten)

___
*[pkg/cmd/mtproxy/main.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBAKnAo38ur45WfqNKW:-MBehwt26072ApaQ7pyV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L52)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

Did you mean `s/InsecureSkipVerify=true/InsecureSkipVerify=false` here?
</blockquote></details>

Done.
___
*[pkg/cmd/mtproxy/main.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBAKvKy6Kut7eQTWDUg:-MBej9YW1A9qOOou2979:b-8xxmet) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L69)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

Setting `options.verify=true` won't work here as it requires a `ServerName` to be specified. I suppose in the long-term, we would want to set `options.verify=true` in production.
</blockquote></details>

It works just fine with IP addresses, but I guess what you are telling me is that in practice we will want to connect via DNS names here, right? In that case, true, we need to set ServerName. I added code to do this in `Proxy()` but I wouldn't be surprised if we found something we need to fix once we try this with "real" certs (which I hope we do soon!)
___
*[pkg/cmd/mtproxy/main.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3nJFd1X9P8DKSlvyl:-MBekgMOBVQ8xVvrk0NL:bn7tmf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L70)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure I like passing the proxying logic as a callback. It might be better to define an interface in `sqlproxy` that is implemented here in some `TenantProxier` struct that is passed in to `Serve`. Makes for better unit testing too I think.
</blockquote></details>

Hmm, yeah, I think it's worth considering this (and your other comments below about reifying some closures). I'm starting to get a little antsy about holding on to this piece of code myself for too long, though. Are you generally on board with leaving the non-functional fixes to this PR as follow-ups in your name? That would free me up to go back to the tenant TLS stuff, which I'm eager to do.
___
*[pkg/cmd/mtproxy/main.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MBALRpG2aTQ6SxLYi96:-MBekzY47BPF6OcYTBmL:bycws51) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/cmd/mtproxy/main.go#L75)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> 			// TODO(asubiotto): implement the actual translation here once it is clear
> 			// how this will work. It's likely that a filename will be passed to the
> 			// proxy which contains a lookup map (and which needs to be re-read on
> 			// SIGHUP). For now, just send everybody to one address and don't validate
> 			// any parameters.
> ```

We can talk about this when you're back, but it would be nice to implement custom routing policies here. I'm looking for a substitution-based routing policy for free tier: https://github.com/imjching/cockroach/commit/545cf62ad7887546c3a7c4489221d07faccef03e#diff-461d7bf3b3646351a9bcb07c0e43b9b5R109-R110. Since the actual translation (i.e. parsing tenant name) is free-tier specific, I wonder if we should allow users to customize what gets extracted. Alternatively, we could store the tenant name in `systems.tenant` as a CRDB thing, so the parsing logic here would make sense.

One of the reasons why I chose routing-based on DNS names is that it reduces a lot of operational work. Using a map here means we need to keep the map up-to-date whenever we modify the tenant list.
</blockquote></details>

Sure! I didn't mean to be prescriptive here, just to jot down the best I could tell this would work. I simply removed this whole comment section and left only a more general TODO.

I didn't think DNS routing would work with all of the pre-warming etc we anticipate we have to do, but that's definitely not my area of expertise. If we can just do that it seems a lot simpler.
___
*[pkg/sql/sqlproxy/proxy.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3q3pK6yINmY1yHyOU:-MBenINiAVNO7YAfvkg7:bse0lga) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I'm not sure either. I thought maybe it had something to do with mandating initialization style, but that doesn't appear to be the case: https://play.golang.org/p/MbfbbR-2uHW.
</blockquote></details>

Huh. I saw this somewhere and copied it. Weird, Goland does complain:

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/T-7WXEIHfjWSoMQeVmGD0t4i/image.png)

Anyway, not worth keeping, removed.
___
*[pkg/sql/sqlproxy/proxy.go, line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qwku41Hp2z2VbVpd:-MBen_1s6uIATkGcCRbs:baininr) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L37)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we move this out of `Proxy`?
</blockquote></details>

I don't feel strongly, mind if I leave it to you to decide in a follow-up? It will require a bit of a yak shave since `sendErrToClient` closes over `opts`.
___
*[pkg/sql/sqlproxy/proxy.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3rhFZ5-UefutOCHIx:-MBenn4NAGyLoAYZoOyG:b-ax8gu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L49)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would extract this block into a `initSecureConn` (or similarly named) function
</blockquote></details>

I think this would make sense if `Proxy` sat on a receiver- as is we'll have to pass lots of random stuff to it. Similar to your previous suggestion, I would leave looking into this to you if you don't mind.
___
*[pkg/sql/sqlproxy/proxy.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t73c7IyWJj1wahOr:-MBeoS_SErQcvwFx-GiO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L60)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Extract `S` into const and reuse below.
</blockquote></details>

Done.
___
*[pkg/sql/sqlproxy/proxy.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t29T0ml7Mo79Sh_h:-MBeotRs-LdiBAhcZNv1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L109)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1. Plus we save the allocations if we do so.
</blockquote></details>

Done.
___
*[pkg/sql/sqlproxy/proxy.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB4r6Fk5dl4Bvf8J1hp:-MBep2tRDMrhH8YJgQWx:bmg7y04) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L128)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

I'm curious, why would this be the case? The goroutines will only run once each, and therefore we will only write to the channels once each.
</blockquote></details>

Each channel will be written to only once, but we only consume from *one* of the channels, so I think @nvanbenschoten is right that we will leak goroutines here.

I will verify that leaktest catches this and then fix... yep it does:

```
=== RUN   TestProxyAgainstSecureCRDB
2020/07/07 21:59:21 handling client 127.0.0.1:50275
2020/07/07 21:59:22 client 127.0.0.1:50275 disconnected after 0.09s: <nil>
    TestProxyAgainstSecureCRDB: leaktest.go:121: Leaked goroutine: goroutine 34 [chan send]:
        github.com/cockroachdb/cockroach/pkg/sql/sqlproxy.Proxy.func3(0x17cdda0, 0xc000394380, 0x17cdda0, 0xc000129180, 0xc000398300)
        	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlproxy/proxy.go:139 +0x104
        created by github.com/cockroachdb/cockroach/pkg/sql/sqlproxy.Proxy
        	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlproxy/proxy.go:137 +0xf49
--- FAIL: TestProxyAgainstSecureCRDB (5.11s)
```
___
*[pkg/sql/sqlproxy/proxy_test.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3wfOb8SfJgnwdZFRS:-MBeyBvaAubNDSNdylYB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L88)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

call `t.Helper` here?
</blockquote></details>

Done.
___
*[pkg/sql/sqlproxy/proxy_test.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3wUCR22uSA_JISZn1:-MBeyCD47u5avqxS4Adr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L88)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

expCode is unused
</blockquote></details>

Done.
___
*[pkg/sql/sqlproxy/proxy_test.go, line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qNSb5IWzccDI24qn:-MBeyEQk4B6lMoHu43g5:b-gw3c3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L178)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is that now?
</blockquote></details>

Yes, but I don't think this PR should be blocked on it as there's some urgency to get the code into mainline. Would you be able to own the TODO? 
___
*[pkg/sql/sqlproxy/server.go, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3us_p5l87XE-HYjJr:-MBeyxEP6hqi8qHD_mV7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/server.go#L10)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like nowhere. They can likely be removed.
</blockquote></details>

Done.
___
*[pkg/storage/enginepb/mvcc3.pb.go, line  at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3uLfJ8XfY0QlQ3Df4:-MBeztXN3TZ254nlHx_x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/storage/enginepb/mvcc3.pb.go#L)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is related to the go mod/1.14 bump, right? Don't think this should be checked in.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented on go.sum:
```diff
@@ -815,6 +915,6 @@ gotest.tools v2.2.0+incompatible/go.mod h1:DsYFclhRJ6vuDpmuTbkuFWG+y2sxOXAzmJt81
-honnef.co/go/tools v0.0.0-20190530104931-1f0868a609b7 h1:RFEMAhc9/Ej5KcaFZooS1PGYAhVl+CUxVj2gyKYGxKQ=
-honnef.co/go/tools v0.0.0-20190530104931-1f0868a609b7/go.mod h1:rf3lG4BRIbNafJWhAfAdb/ePZxsR/4RtNHQocxwk9r4=
+honnef.co/go/tools v0.0.1-2019.2.3 h1:3JgtbtFHMiCmsznwGVTUWbgGov+pVqnlf1dEJTNAXeM=
+honnef.co/go/tools v0.0.1-2019.2.3/go.mod h1:a3bituU0lyd329TUQxRnasdCoJDkEUEAqEt0JzvZhAg=
```
last time we upgraded honnef.co like this, StaticCheck started flaking (seemed like a honnef.co bug).

I forced this back by doing `go get honnef.co/go/tools/cmd/staticcheck@1f0868a6`.

Upgrading may also be a way forward but there were multiple lint errors triggered, so I avoided that step temporarily.

otan commented on go.mod:
```diff
@@ -89,7 +88,9 @@ require (
 	github.com/jackc/fake v0.0.0-20150926172116-812a484cc733 // indirect
+	github.com/jackc/pgproto3/v2 v2.0.2
 	github.com/jackc/pgx v3.6.2+incompatible
+	github.com/jackc/pgx/v4 v4.7.1
```
this won't be an issue -- older pgx imports will import the old library (as go.mod intended).

imjching reviewed (commented):
Thanks for filing the issue. 
Overall approach seems fine to me. I'll leave the rest of the reviews to others. :)

> Reviewed 2 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MBiMjDT-YVYQsdbdBv7:b-50reu1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @imjching, @nvanbenschoten, and @tbg)

___
*[pkg/sql/sqlproxy/proxy.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t29T0ml7Mo79Sh_h:-MBiL5ZVE63onWMRMEFd:b-7vd1qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L109)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

We need to use `pgSSLRequest` here.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> let's get this in

> Reviewed 36 of 36 files at r3, 6 of 6 files at r4, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MBmtxuV5xw7nVvd-2oH:b-2tsqkr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @tbg)

___
*[pkg/sql/sqlproxy/proxy_test.go, line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3qNSb5IWzccDI24qn:-MBmthm71dbR9zolPpZ7:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy_test.go#L178)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Yes, but I don't think this PR should be blocked on it as there's some urgency to get the code into mainline. Would you be able to own the TODO? 
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTRs! I moved the package to ./pkg/ccl/sqlproxyccl and updated the licenses as well as made minor edits to the commit. Will merge when I've gotten this to pass CI.

> Dismissed @asubiotto and @otan from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-:-MBniqVH3hmUWTTJcYTk:byvmpif)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @imjching, and @nvanbenschoten)

___
*[pkg/sql/sqlproxy/proxy.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50647#-MB3t29T0ml7Mo79Sh_h:-MBneSRW7v8SW2sSi3jW:bmoi04s) ([raw file](https://github.com/cockroachdb/cockroach/blob/624f551271c1c0572bd7c0efc0dc73d6c7427acf/pkg/sql/sqlproxy/proxy.go#L109)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

We need to use `pgSSLRequest` here.
</blockquote></details>

:facepalm: done now.


<!-- Sent from Reviewable.io -->

tbg committed:
sqlproxy: initial commit and end-to-end demo

This adds the proxy envisioned for use with multi-tenant deployments.
Initially, the proxy routes all incoming connections to a backend SQL
server indicated via a command-line flag.

This is good enough for initial sanity checks, but the following changes
need to be made:

- allow `--insecure` (for testing purposes only), which means that some
  of the steps in the proxy are skipped
- add an in-memory Proxy() to (TestServer).StartTenant as well as the
  3node-tenant logic test configuration.
- move `pkg/cmd/mtproxy` to `./cockroach mt start-proxy`. In doing so,
  the proxy will get proper certificate management (the certs types
  will have to be added though), which makes it easier to use.
- implement the parsing of tenant identifier from the database name or
   options field
- load a tenant identifier -> tenant ID map from a file specified via a
    command-line flag
- reload that map on SIGHUP
- unit testing of the resulting code used in the `./cockroach mt
  start-proxy` command (there is already some testing of the `sqlproxy`
  package itself)

The following snippet sets up a demo SQL shell connecting, securely,
through the proxy to the SQL tenant server, which in turn talks to
the KV layer.

    #!/bin/bash

    set -euxo pipefail

    # Prep work
    rm -rf ~/.cockroach-certs cockroach-data
    killall -9 cockroach || true
    killall -9 mtproxy || true

    export CERTSDIR=$HOME/.cockroach-certs
    export COCKROACH_CA_KEY=$CERTSDIR/ca.key
    ./cockroach cert create-ca
    ./cockroach cert create-client root
    ./cockroach cert create-node 127.0.0.1 localhost

    #  Start KV layer. (:26257)
    ./cockroach start --host 127.0.0.1 --background
    # Create tenant
    ./cockroach sql --host 127.0.0.1 -e 'select crdb_internal.create_tenant(123);'

    # Spawn tenant SQL server (:36257) pointing at KV layer (:26257)
    ./cockroach mt start-sql --tenant-id 123 --kv-addrs 127.0.0.1:26257 --sql-addr 127.0.0.1:36257 &
    sleep 1

    # Create user on tenant (proxy does not forward client certs)
    ./cockroach sql --port 36257 -e "create user bob with password 'builder';"

    # Spawn the proxy on :5432, forwarding to tenant SQL server (:36257)
    go run ./pkg/cmd/mtproxy/ -cert-file $CERTSDIR/node.crt -key-file $CERTSDIR/node.key -verify=false --target 127.0.0.1:36257 &

    sleep 5

    # Connect to proxy.
    ./cockroach sql --host 127.0.0.1 --port 5432 --user bob

Release note: None

tbg commented:
CI failed on a known panic in TestLogic, so going ahead with a merge attempt

bors r=asubiotto
